modules/diskq/logqueue-disk-non-reliable.c

```diff
@@ -649,6 +651,139 @@ _push_tail(LogQueue *s, LogMessage *msg, const LogPathOptions *path_options)
     scratch_buffers_reclaim_marked(marker);
 }
 
+static void
+_push_tail(LogQueue *s, LogMessage *msg, const LogPathOptions *path_options)
+{
+  LogQueueDiskNonReliable *self = (LogQueueDiskNonReliable *) s;
+  gint thread_index;
+  LogMessageQueueNode *node;
+
+  thread_index = main_loop_worker_get_thread_index();
+
+  /* if this thread has an ID than the number of input queues we have (due
+   * to a config change), handle the load via the slow path */
+
+  if (thread_index >= self->num_input_queues)
+    thread_index = -1;
+
+  /* NOTE: we don't use high-water marks for now, as log_fetch_limit
+   * limits the number of items placed on the per-thread input queue
+   * anyway, and any sane number decreased the performance measurably.
+   *
+   * This means that per-thread input queues contain _all_ items that
+   * a single poll iteration produces. And once the reader is finished
+   * (either because the input is depleted or because of
+   * log_fetch_limit / window_size) the whole bunch is shared between the
+   * front_cache and legacy behaviour.
+   */
+
+  if (thread_index >= 0)
+    {
+      /* fastpath, use per-thread input FIFOs */
+      if (!self->input_queues[thread_index].finish_cb_registered)
+        {
+          /* this is the first item in the input FIFO, register a finish
+           * callback to make sure it gets moved to the wait_queue if the
+           * input thread finishes
+           * One reference should be held, while the callback is registered
+           * avoiding use-after-free situation
+           */
+
+          main_loop_worker_register_batch_callback(&self->input_queues[thread_index].cb);
+          self->input_queues[thread_index].finish_cb_registered = TRUE;
+          log_queue_ref(&self->super.super);
+        }
+
+      log_msg_write_protect(msg);
+      node = log_msg_alloc_queue_node(msg, path_options);
+      iv_list_add_tail(&node->list, &self->input_queues[thread_index].items);
+
+      self->input_queues[thread_index].len++;
+
+      log_msg_unref(msg);
+      return;
+    }
+  // slow path
+  _push_tail_single_message(s, msg, path_options);
+}
+
+static void
+_move_part_of_input_to_front_cache(LogQueueDiskNonReliable *self, InputQueue *input_queue,
+                                   gint num_msgs_to_send_to_front_cache)
+{
+  struct iv_list_head *head = &input_queue->items;
+  struct iv_list_head *pos = head->next;
+  gint count = 0;
+  g_mutex_lock(&self->super.super.lock);
+  while (pos != head && count < num_msgs_to_send_to_front_cache)
+    {
+      struct iv_list_head *next = pos->next;
+      LogMessageQueueNode *node = iv_list_entry(pos, LogMessageQueueNode, list);
+      LogPathOptions lpo = LOG_PATH_OPTIONS_INIT;
+      lpo.ack_needed = node->ack_needed;
+      lpo.flow_control_requested = node->flow_control_requested;
```

Comment:
We do not need to copy the flow_control_requested bit here. 

That's only used while the message is traversing the LogPipe pipeline. By the time we are in the queue, the flow_control_requested bit turns into an ack_needed TRUE/FALSE, see the comment on the flow_control_requested  member of LogPathOptions in logpipe.h

If the comment is not enough there at LogPathOptions to make this clear, please improve it.